### ISSUE TITLE
Timebox: Proof of concept for real reference counting in Flux engine

### ISSUE THREAD

**Issue by wolffcm** [influxdata/flux#3865]
We would like to improve the efficiency of the Flux engine by reducing the amount of memory needed and number of allocations performed. One way to do this is to perform reference counting on tables and or columns (arrow arrays) within those tables.

For an example at the table level, in an example like this:
```js
tables = from(..) |> range(..) |> filter(...)
tables |> sum() |> yield(..)
tables |> count() |> yield(..)
tables |> min() |> yield(..)
tables |> max() |> yield(..)
```
The tables in the stream `tables` need to be consumed by 4 different successor transformations. Rather than copy all of those tables multiple times as we do today, it would be more efficient to share single copies of these read-only tables among all 4 transformations.

DOD:
- Come up with a proof of concept for avoiding copying tables and/or columns when not needed.
  - (The POC does not need to cover all possible cases, and is assumed to be throwaway code.)
- Show that for some use case that allocations and/or GC activity is reduced and as a result throughput increases
- Come up with a strategy for landing an optimization like this in production (another working group might be appropriate) 

NOTE:

At the end of the timebox we will close this issue and create more specific issues for any outstanding question/design concerns.

**Comment by jsternberg** [comment]
There's two components to this issue that will need to be evaluated separately and it might be good if we split them up. I'm going to address them in reverse order because the first will probably involve more design and unknowns. I'll give a quick summary in this comment and then try to write up more detailed comments on each of these two.

## Multiple Successor Transformations

> For another example at the table level, in an example like this:
>
>     tables = from(..) |> range(..) |> filter(...)
>     tables |> sum() |> yield(..)
>     tables |> count() |> yield(..)
>     tables |> min() |> yield(..)
>     tables |> max() |> yield(..)
>
> The tables in the stream tables need to be consumed by 4 different successor transformations. Rather than copy all of those tables multiple times as we do today, it would be more efficient to share single copies of these read-only tables among all 4 transformations.

At the present moment, this is exactly how those tables are processed. If you wrote this query, it would _not_ create 4 copies of the tables that are consumed by `sum()` and the other aggregates. It would create one copy of the data for all of those.

What it would do is it would be greedy and reads the entire output of `filter()` into memory before passing it to the successor transformations. This is because table streams are presently not capable of being interrupted during processing so it is not possible to process part of a table and then come back to process the rest.

## Analyzing Flux Functions

> Consider for example that the output of many transformations might leave an input column untouched. For example, in this very common pattern:
>
>     ...
>       |> map(fn: (r) => ({r with newCol: r._value * 10}))
>     ...

This one is interesting because I think it touches a similar place as vectorization. In this case, the map function has some semantic graph and it needs to determine the meaning of the semantic graph. It needs to determine which values are accessed, transformed, and ultimately output. At the current moment, our interpreter runtime (in code, it's called the compiler and is in the compiler package) is very naive about this and ultimately uses a lot of memory to capture what is happening. The `{}` creates a new object and allocates memory each time that the function is executed. The `map()` function suffers from this more because object creation is a lot of small allocations. In general, the functions passed to filter end up being much faster because they don't allocate as much memory.

There's a few major components to this:

1. Code evaluation should be more efficient. We have enough information that we can properly compile this statement rather than interpreting it from semantic graph.
2. Memory allocation does not need to be done for each row. We know what `r` contains for the processing of a table with a specific group key and we don't need to repeat this allocation for every row. We can reuse the same memory within the runtime stack. I'm not referring to arrow arrays here. I'm referring to the equivalent of registers/stack memory.
3. I think we'll need to probably design an IR language to capture program flow. If we rely on processing semantic graph, we're going to miss a lot of optimization opportunities. This is the kind of analysis that SSA IR was designed for. For example, the following is equivalent to the above:

```
myfunc = (r) => {
    val = r._value * 10
    return {r with newCol: val}
}
```

But if we relied solely on the present semantic graph, I think we'd start to have difficulty. Instead, I think we need to consider transforming this into an SSA IR. The above two would convert into the same IR.

```
define record @myfunc(record %r) {
    %1 = getvalue %r, "_value"
    %2 = loadconst 10
    %3 = mul %1, %2
    %4 = extend %r, "newCol", %3
    ret %4
}
```

While SSA IR's are traditionally output as assembly, they naturally convert directly into the equivalent of our planning graph. I suspect this will open up more opportunities for modifying the IR for optimization. It also has a few other side benefits:

1. It's no longer possible for a person to break the stack by having 1,000 conditions since processing the function will no longer involve any recursion.
2. Recursion isn't needed for analyzing or modifying the function.

This is very abbreviated because I'm just trying to get some thoughts down. I'll try to expand more. I might have to do some digging in to see if my thoughts lead anywhere useful.

**Comment by jsternberg** [comment]
## Multiple Successor Transformations

As mentioned above, we are limited in our ability to parallelize and use the existing pipeline mechanics with our current engine. Our engine is designed as a pipeline similar to as follows:

```
A -> B -> C
```

This is where `A` is a source and `B` and `C` are transformations. The pipeline is enabled by the dispatcher which decides which node of the pipeline is going to run at any given moment. Incoming messages to a transformation node will be pushed into the message queue and tell the dispatcher that it is ready to process new messages. The dispatcher uses a FIFO algorithm to determine which node to run. The dispatcher will limit a node to processing 10 messages (the default) before putting it back at the end of the queue and processing more messages.

The dispatcher may be interrupted and it will stop processing messages if it is. An interrupt may happen because some node returns an error or the context gets canceled due to a query being canceled.

The dispatcher also has the capability to run multiple nodes concurrently. This is not used frequently as it isn't configurable and the current model depends on the number of leaf nodes (successors) in the graph. If there is only one yield, then the concurrency is one.

An important distinction: **throughput** is how many messages a node will process and **concurrency** is how many nodes may process messages at the same time. Additionally, a node may only be run in one worker at a time. It is not possible for the same filter node to run concurrently with a table from another group key unless goroutines are used to break the existing contract. There have been existing bugs because we used goroutines to break this contract that have caused race conditions.

Each node above **cannot be run concurrently in multiple dispatcher threads**. If you up the number of 

The underlying algorithm described above doesn't really have many glaring problems and I'm not sure it has to be changed significantly or at all. The above algorithm is solid. What's not solid is the messages that presently exist and how they are processed by transformations.

First, 

A current weakness of our pipeline is that **sources do not participate in the pipeline**. They spawn their own goroutines and continuously feed messages into their successor transformation. They do not have a

**Comment by jsternberg** [comment]
I accidentally pressed close with comment while typing a comment. Finishing what I typed and deleted the above unfinished comment.

**Comment by wolffcm** [comment]
I've been thinking about this. I think it would be possible to be able to do the analysis we need  (as to which values are accessed, transformed etc), leverage vectorization, and still use our existing semantic graph as the representation. 

Consider a call to map with a lambda, where the fields of `r` are all floats:
```
map(fn: (r) => ({_value: r.a * r.b + r.c}))
```
With vectorization, we want to be able to deal with fields of `r` as vectors, rather than individual values. We could transform this lambda with a `Vectorization` code transformation like this:
```
(_r_a, _r_b, _r_c) => ({_value: _r_a * _r_b + _r_c})
```
Where the types of `_r_a` etc are `[float]`, and the return type is `{_value: [float]}`. 

The steps for this code transformation would be something like this:
1. We start knowing the actual concrete type of `r` and all its fields
2. Replace any instances of `{r with ...}` with a record expression that explicitly includes all fields
3. Replace the `r` parameter with the individual parameters for each field that is referenced via a `MemberExpression`. 
4. Replace each `MemberExpression` that uses `r` with an `IdentifierExpression` like `_r_a`, etc.

Then we could evaluate the lambda by calling it with arrays exposed by the `ColReader`.

By exploding the fields of the record out into individual arrays, we get closer to the vector operations we actually want to perform, and it also becomes easier to analyze how each array input is used (we might consider implementing a rudimentary version of [use def chains](https://en.wikipedia.org/wiki/Use-define_chain)). Note that semantic graph is already in SSA form; variables can only be assigned once statically.

We would need to make the compiler able to support elementwise vector operations. And we may wish to extend the semantic graph representation itself to make elementwise operations explicit, by adding new expression types like `BinaryElementwiseExpression`, `CallElementwiseExpression` and so on.

I don't see a whole lot of value in a three-address-code style IR except where we are concerned about stack depth and recursion. But since Go can grow the stack dynamically, and we have alternatives like iterative traversal using stacks or queues allocated on the heap, it seems a minor concern, unless I'm missing something.

**Comment by jsternberg** [comment]
@wolffcm I'm going to create a new issue for the vectorization stuff and we can copy/paste the issue there. I've been investigating the multiple successor transformations and the buffer stuff for this issue. I'd like to treat the timeboxes as separate since the two solutions will end up being pretty different.

**Comment by jsternberg** [comment]
I split the issue into #3877, lowered the estimate for this issue, and added an estimate to the new issue and added it to the current sprint.

**Comment by jsternberg** [comment]
Leaving some initial notes for improving memory usage and throughput.

## Background

Memory usage is the amount of memory used. So unnecessary copying of data is a memory usage problem. Memory throughput is about improving the amount of memory used during the lifetime of the query and the data.

To illustrate the difference, we can look at the `group()` function in flux as it presently is. The `group()` function has very good memory usage as it does not allocate any memory and reuses the existing already allocated memory. It would add nothing to the amount of memory used. Its memory throughput is very poor though as it will continue reading table buffers and it will not release them until it has read them all.

These two concepts get captured by two stats that we capture as part of the query, but these stats don't perfectly encapsulate how memory is used throughout the system. We don't presently have a way to granularly track where and how memory is used and released. That might be something to attempt as part of this timebox.

The `TotalAllocated` statistic captures memory usage. The `group()` method adds nothing to the `TotalAllocated` statistic.

The `MaxAllocated` statistic captures memory throughput. The `group()` method will cause `MaxAllocated` to spike as it will hoard the data before releasing it to the next transformation.

For my initial tests, I adapted and reused my hackday project that was related to improving the memory throughput. I will also be focusing a lot on memory throughput because memory usage is dependent on the transformations. There's really nothing the engine can do to improve the memory usage for the `derivative()` function. This requires rewriting the `derivative()` function. All of the transformations that have good memory usage have gone through rewrites and this will probably continue to be the case.

## Global Changes

I made a few global changes that aren't present on the main/master branch.

I added a special command line runner that would execute a query without going through the REPL. This matches how queries get run inside of influxdb and the cloud environment and how the CLI should run queries from a file. This is to give the most accurate comparison.

I modified the default concurrency for a query to be the number of cpus on my personal machine with `runtime.GOMAXPROCS(0)` instead of defaulting to 1. This is meant to see how well the changes to the engine help with pipeline concurrency. Pipeline concurrency is the ability for the dispatcher to execute multiple transformations at the same time using multiple cores. For production, this value can continue to be 1 or we can start to experiment with a smarter dispatcher/scheduler algorithm for multiple tenancy. As part of this change, we are not testing to see if it would improve query speed although if it slowed down queries we will make note of it. I personally did not see any slowdown or improvement from the tests I have done so far.

The last change was to add opentracing to the command line runner so that I could export the traces to a local jaeger instance to analyze the results. This was primarily done for understanding how and when transformations are executed for understanding pipeline concurrency. I may also try to take advantage of this to better track memory usage and lifetimes.

## Engine Changes

The change to the engine I am testing is whether processing individual buffers rather than whole tables improves memory throughput and pipeline concurrency without being detrimental to performance or general memory usage.

At the current moment, a `flux.Table` will be passed as a single message to the next transformation to process. This change modifies it so a `flux.ColReader` equivalent gets passed as a single message instead and that buffers for tables may occur out of order instead of being forced to process all buffers from one group key and then move to the next one.

## Initial Tests

I performed my initial tests using the given query:

```
import "internal/gen"
import "strings"

gen.tables(
  n: 4096,
  tags: [
    {name: "_measurement", cardinality: 1},
    {name: "_field", cardinality: 1},
    {name: "t0", cardinality: 5},
    {name: "t1", cardinality: 10},
  ]
)
  |> filter(fn: (r) => r._value > 0.0)
  |> map(fn: (r) => ({r with t0_count: strings.strlen(v: r.t0)}))
  |> filter(fn: (r) => r.count > 3)
```

`filter` and `map` were used because those were two of the functions I had modified for the hackday project. Filter uses the same basic code as what is presently in Flux but it removes stream because it would be unnecessary in this system. Map was modified more to use arrow builders and map will now send buffers immediately instead of hoarding the data like group does.

`group` has also been modified to hopefully improve memory throughput but I have not tested it yet.

I tested three versions and retrieved the `maxAllocated`, `totalAllocated`, and made observations on the trace.

```
maxAllocated=1711680 (1.7 MB) totalAllocated=15757056 (15.7 MB)
maxAllocated=11087168 (11 MB) totalAllocated=27132288 (27.1 MB)
maxAllocated=19898624 (19.9 MB) totalAllocated=72005120 (72 MB)
```

The first is new code with the new system. It yielded a lower total allocated which may be due to the map changes that switched it to using arrow directly instead of the column list table builder, but this also resulted in a lower max allocated indicating a better usage of that memory. The percentage difference between max and total is 10.8% which may indicate a more efficient pipeline.

The second modifies the current filter stream to use the buffered builder (so that we don’t have stream’s goroutines interfering) and have a closer comparison to what filter should presently be in the existing system. The current filter breaks rules in a detrimental way and wasn’t suitable for isolated testing. This change resulted in a higher max allocated and higher total allocated, but the trace indicated that data was still moving through the pipeline and there didn’t seem to be a very large issue with concurrency. It did seem like the new engine lowered the maximum allocated memory a fair amount as the percentage difference between the two is 40% which indicates a worse memory throughput since memory could not be released as efficiently. Surprisingly, pipeline concurrency was decent between filter and map. This seems to be because of the special code added to filter with the passthrough dataset which allows data to be processed by the next transformation. More expectantly, the map transformation did not pass data to the second filter until it had already processed all data.

The third modified filter to use column list table builder and removed the narrow transformation hack. This was to give a more accurate comparison to filter before we did some performance improvements on it and give a better idea of how most other existing transformations that did not get rewrites would behave. As seen, total allocated is much higher and max allocated is also higher. This may just be because the column list table builder is not very good at either of these. When looking at the trace for pipeline concurrency, this last one was the one that did not benefit from increasing the concurrency because each node required the previous one to finish before starting. There was no concurrency.

## Further Work

Further work is to test group and an aggregate which should benefit from this change. While the existing changes seem to have improved memory throughput, the new engine should automatically optimize `group |> count` or other aggregates without any special planner rules or code and should pretty drastically improve the memory throughput. I'll also test `filter |> count` and `filter |> group`.

**Comment by jsternberg** [comment]
Here are some numbers for this query. I only tested current (with the above mentioned filter change to remove stream) and the new engine stuff.

```
import "internal/gen"
import "strings"

gen.tables(
  n: 4096 * 10,
  tags: [
    {name: "_measurement", cardinality: 1},
    {name: "_field", cardinality: 1},
    {name: "t0", cardinality: 5},
    {name: "t1", cardinality: 10},
  ]
)
  |> filter(fn: (r) => r._value > 0.0)
  |> group(columns: ["t0"])
  |> count()
```

```
maxAllocated=1896064 (1.9 MB) totalAllocated=6840440 (6.8 MB) (4096 pps, experimental engine)
maxAllocated=3466752 (3.5 MB) totalAllocated=8639992 (8.6 MB) (4096 pps, current engine)
maxAllocated=1727104 (1.7 MB) totalAllocated=51599480 (51.5 MB) (40960 pps, experimental engine)
maxAllocated=18940672 (18.9 MB) totalAllocated=70884728 (70.8 MB) (40960 pps, current engine)
```

pps is points per series. I tested with the same cardinality but more data as mentioned above. For the current engine, pipeline concurrency crashed after the group call and all aggregate calls were invoked together. For overall speed, this didn't really matter. For the experimental engine, it was able to process the result of group as it came through. As seen, the experimental engine has much better memory throughput as the max allocated is much smaller with the increased amount of data.

An implementation detail I found in the experimental engine is that the tracing is too granular at the moment. With the increase in the number of messages, opentracing created too many traces. This is an implementation problem and I don't think it's inherent to this model. I think can move the opentracing spans a bit to get the same granularity without the trace becoming too big. It didn't seem to impact performance very much. The query was still around the same amount of time.

It does appear like some of the queries are consistently slower by a small percentage (like 900 ms to 1 second or 93 ms to 110 ms). I'm not sure if this is because of the engine or because of the transformation changes. In particular, the filter in the experimental engine is doing more work because it prepares the function per buffer instead of per table. I think we can optimize this to cache the prepared function instead of redoing it per buffer. Redoing it per buffer was just easier to write.

**Comment by jsternberg** [comment]
After some additional testing I found that the lack of updates to aggregates to work with the new engine did skew the results to have fewer gains than I think they should. I'll make those changes and see if the results are different.

**Comment by jsternberg** [comment]
I made the mentioned change to count and compared this query:

```
import "internal/gen"
import "strings"

gen.tables(
  n: 4096 * 10,
  tags: [
    {name: "_measurement", cardinality: 1},
    {name: "_field", cardinality: 1},
    {name: "t0", cardinality: 5},
    {name: "t1", cardinality: 10},
  ]
)
  |> group(columns: ["t0"])
  |> count()
```

This resulted in a pretty drastic difference in memory throughput.

```
maxAllocated=241920 (0.24 MB) totalAllocated=51804160 (51.8 MB)
maxAllocated=35330560 (35 MB) totalAllocated=51717240 (51.7 MB)
```

The first is with the engine changes and the second is base flux.

Some pretty key observations I have.

1. The memory throughput is generally better when transformations are modified to take advantage of them. Since I am working based on my dev friendly flux hackday project, the transformations are generally easier to write too.
2. Speed has been consistently a bit slower. There's two possibilities for this. First is that the new engine results in more messages being created and consumed. This can cause more memory to be used for the message queues or it could be because the opentracing that's presently implemented as part of this is creating too many spans. I don't think the message queues are the problem since the traces seem to indicate that messages are being consumed pretty quickly and that makes me think it's unlikely to be the problem. I think opentracing is probably more likely so it might be a thing to watch out for if we productionize this branch. It's also possible some transformations, like filter and map, are doing more work than necessary because of poor coding.

**Comment by jsternberg** [comment]
I'm attempting to observe memory profiles and when I use the `group |> count` example I don't even see them show up on the memory profile. It doesn't appear like there's a negative effect on non-tracked portions of the engine so the improvements to tracked memory should generally improve memory performance.

**Comment by jsternberg** [comment]
Issues have been created for this so I'm going to close this issue.
